cspreports/tests/test_utils.py

```diff
@@ -191,3 +170,88 @@ def test_naive(self):
             mock_now = datetime(2016, 4, 27, 12, 34)
             with patch('cspreports.utils.now', return_value=mock_now):
                 self.assertEqual(get_midnight(), datetime(2016, 4, 27))
+
+
+class SaveReportTest(TestCase):
+
+    def test_save_report_missing_root_element(self):
+        """ Test that the `save_report` handler correctly saves to the DB. """
+        assert CSPReport.objects.count() == 0  # sanity
+        body = '{"document-uri": "http://example.com/"}'
+        request = RequestFactory(HTTP_USER_AGENT='Agent007').post('/dummy/', body, content_type=JSON_CONTENT_TYPE)
+
+        utils.save_report(request)
+
+        reports = CSPReport.objects.all()
+        self.assertQuerysetEqual(reports.values_list('user_agent'), [('Agent007', )], transform=tuple)
+        report = reports[0]
+        self.assertEqual(report.json, body)
+        self.assertFalse(report.is_valid)
+
+    def test_save_report_no_agent(self):
+        """Test that the `save_report` handler correctly handles missing user agent header."""
+        request = RequestFactory().post('/dummy/', '{"document-uri": "http://example.com/"}',
+                                        content_type=JSON_CONTENT_TYPE)
+
+        utils.save_report(request)
+
+        reports = CSPReport.objects.all()
+        report = reports[0]
+        self.assertQuerysetEqual(report.user_agent, '')
```

Comment:
```diff
-        report = reports[0]
-        self.assertQuerysetEqual(report.user_agent, '')
+        self.assertQuerysetEqual(reports.values_list('user_agent', flat=True), [''])
```

Check both the number of reports and the value.

Reply:
I've noticed that the suggested implementation is failing on some python versions. I've changed to:
```python
        report = CSPReport.objects.first()
        self.assertQuerysetEqual(report.user_agent, '')
```

hope that works anyway

Reply:
I don't remember having problems with this construct, but OK.